tests/unit/test_charm.py

```diff
@@ -0,0 +1,335 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+from pathlib import Path
+from unittest.mock import PropertyMock, patch
+
+import pytest
+import yaml
+from ops.model import BlockedStatus, WaitingStatus
+from ops.testing import Harness
+from tenacity.wait import wait_none
+
+from charm import KafkaCharm
+from literals import CHARM_KEY, PEER, REL_NAME, ZK
+
+logger = logging.getLogger(__name__)
+
+
+CONFIG = str(yaml.safe_load(Path("./config.yaml").read_text()))
+ACTIONS = str(yaml.safe_load(Path("./actions.yaml").read_text()))
+METADATA = str(yaml.safe_load(Path("./metadata.yaml").read_text()))
+
+
+@pytest.fixture
+def harness():
+    harness = Harness(KafkaCharm, meta=METADATA)
+    harness.add_relation("restart", CHARM_KEY)
+    harness._update_config(
+        {
+            "offsets-retention-minutes": 10080,
+            "log-retention-hours": 168,
+            "auto-create-topics": False,
+        }
+    )
+    harness.begin()
+    return harness
+
+
+def test_install_sets_opts(harness):
+    with patch("snap.KafkaSnap.install"), patch(
+        "config.KafkaConfig.set_kafka_opts"
+    ) as patched_opts:
+        harness.charm.on.install.emit()
+
+        patched_opts.assert_called_once()
+
+
+def test_install_waits_until_zookeeper_relation(harness):
+    with patch("snap.KafkaSnap.install"), patch("config.KafkaConfig.set_kafka_opts"):
+        harness.charm.on.install.emit()
+        assert isinstance(harness.charm.unit.status, WaitingStatus)
+
+
+def test_install_blocks_snap_install_failure(harness):
+    with patch("snap.KafkaSnap.install", return_value=False), patch(
+        "config.KafkaConfig.set_kafka_opts"
+    ):
+        harness.charm.on.install.emit()
+        assert isinstance(harness.charm.unit.status, BlockedStatus)
+
+
+def test_leader_elected_sets_passwords(harness):
```

Comment:
might be nice to check other cases of leader elected

Reply:
What do you mean by other cases? I think the only thing it currently does is set this relation data that we're checking for. Are there other cases you're thinking of?